utils/array.js

```diff
@@ -1,8 +1,8 @@
-import { xor } from 'lodash';
+import { xor, isEqual } from 'lodash';
 import { get } from '@/utils/object';
 
 export function removeObject(ary, obj) {
-  const idx = ary.findIndex(resource => resource.id === obj.id);
+  const idx = ary.findIndex(item => isEqual(item, obj));
```

Comment:
Still no.. `removeObject` is not wrong here, it is meant to remove a particular exact entry, not to look for an equivalent one (this will also be thousands of times slower in some cases).

The root problem that you have equivalent-but-not-equal entries in the first place needs to be addressed (probably in steve client)

Reply:
I think I see the underlying problem. The socket resource.remove listener uses getters.byId which uses the type cache's map to find reference to the object to be removed, then dispatches the remove mutation, which attempts to remove that object from the cache's list and map. The list removal doesn't work sometimes (this is causing the need-refresh-to-see-delete bug) because the list and map refer to different but equal objects sometimes.

The delete bug doesn't appear if the page is refreshed before deleting any new resources, which makes me think the problem is in the load mutation. When a resource is created in-ui, both the resource-instance save function and the socket resource.create listener trigger the load mutation.

The former includes an 'existing' proxy object, which, for new resources, was instantiated in ResourceDetail when the page loaded. If the socket load mutation fires before the resource-instance.save() mutation, the list and map both get the updated/created resource from the first mutation, but then with the second mutation the type cache's map gets updated with that 'existing' proxy object, and the map and list are no longer in sync.

If I'm right about the bug, I think of a few ways of addressing it to varying extent:
        * don't call load mutations from both save() and ws.resource.create
        * don't pass an existing proxy to load mutation ever
        * don't pass an existing proxy object to load mutation unless that existing proxy is already in the store (edit operations)
        * don't use the existing proxy object passed to load unless this is an update operation (check that there is already a proxy for this resource in store and ignore the 'existing' argument)
        * ~~don't require the list and map to have the same object references~~
        * don't track both list and map in vuex (make one a getter off the other?) so only one needs to be updated in load and delete

Ultimately, I think maintaining both a list and map of the same information is introducing an unnecessary degree of fragility. If the problem with my previously suggested solution is that it is insensitive to equal-but-separate entries in list and map then we should probably also change the code that allows equal but separate entries to occur at all.

Reply:
@codyrancher This might be related to your problem too or at least be interesting.

> Ultimately, I think maintaining both a list and map of the same information is introducing an unnecessary degree of fragility.

This is fundamentally necessary for reasonable performance.  There are 2 basic access patterns:
  - Give me an array of all the resources of one type (to show list pages, the live "all" arrays of a type, etc).
    - Storing only a map just give you no good way to do this.
  - Give me an individual entry by type/id (or tell me if one exists)
     - Storing only an array and using findIndex() would be many times slower than a map lookup; it scales at-best linearly with the number of resources.
    - We need to be able to handle tens of thousands+ of a single kind of resource being loaded.
    - findIndex on an array of 10,000 things to find an ID that doesn't exist is 10,000 loops and 10,000 invocations to the testing function.  This gets ugly real fast (and is the way the store for the old UI worked originally a long time ago...)

> the list and map both get the updated/created resource from the first mutation, but then with the second mutation the type cache's map gets updated with that 'existing' proxy object, and the map and list are no longer in sync.

Unless something else is wrong, this is not a problem on its own.  The cache and list item are (supposed to be) two references to the **same** single object instance, so updating one of them updates "both".  Other than them being stored in maps and arrays, this is logically the same as:

```
x = { a: 1 };
y = x;
console.log(x.a); // 1
console.log(y.a); // 1

x.a = 2;
console.log(x.a); // 2
console.log(y.a); // 2
```

> If the problem with my previously suggested solution is that it is insensitive to equal-but-separate entries in list and map then we should probably also change the code that allows equal but separate entries to occur at all.

Yes, if this is true (not just 2 references to the same instance) then this is definitely the root problem that needs to be fixed.  For a given type/id there should _always be exactly one_ instance of the model/proxy in the store for it.  This is critical to how that change events can update it and trigger other places that use it to update efficiently.

The only time there is (should be) separate instances of a single resource is when you're explicitly `clone`-ing it.  We do this on e.g. edit screens so you're editing a stable set of values; on save those changes are pushed back into the main copy in the store and the clone is forgotten.

Create is a special case of that, and that's what the `existing` input to the mutation is for.  On create your code might already have a handle to an instance of the new resource which you want to keep.  e.g. a multi-step create that does something to the effect of `a = new A(...); a.save(); B.foo = A.id, B.save()`.  So for the create case, the model on which save() is being called on becomes the copy that is put into the store, instead of the load() mutation making a new instance and saving that.

That part is relatively new, so it's entirely possible it's broken in some case.. before then the original `A` would be an orphaned model unconnected to the one now persisted in the store (so e.g. it would never get its `id` populated).  The only reason a representation of `A` ever got into the store at all before was because a change event came in and created one shortly after.  That change event should be _updating_ the instance that has already been put into the store (probably with no actual change, since the save API call returned the same representation).

Reply:
> Give me an array of all the resources of one type (to show list pages, the live "all" arrays of a type, etc).
Storing only a map just give you no good way to do this.

What? Maps are iterable, efficiently iterable not like a for...in on an object or Object.values() or whatever.

>Unless something else is wrong, this is not a problem on its own. 

I don't think I explained myself clearly enough. This is in the load mutation:

```
load(state, { data, ctx, existing }) {
    let type = normalizeType(data.type);
    const keyField = KEY_FIELD_FOR[type] || KEY_FIELD_FOR['default'];

    const id = data[keyField];

    let cache = registerType(state, type);

    let entry = cache.map.get(id);

    if ( existing ) {
      // A specific proxy instance to used was passed in (for create -> save),
      // use it instead of making a new proxy
      entry = existing;
      Object.assign(entry, data);
      cache.map.set(id, entry);
    } else if ( entry ) {
...blah blah the rest works
```
The new entry isn't getting added to the list. We can't just push() the new entry in because this block of code also runs on editing operations, we have to use addObject() to check if its already there and update it if it is, but addObject wont account for the possibility that a different proxy object with the same data is already in the list. There's a fair chance that the resource already has a different proxy in the list because both the socket and the resource-instance save() function are trying to add one around the same time.

Of course we can retool the load mutation/how it's called to fix this specific problem, but I still think having to maintain two sources of truth introduces a lot of opportunity for error.

Reply:
`existing` is only used in one place: resource-instance save().  Yes, it appears the `addObject(cache.list, entry)` is missing that case and should be added (like I said, the `existing` stuff is pretty new).  So that's one bug, and that will cause new items to not show up on the list page after save until you reload (which I think is your thing, @codyrancher ?)

Switching everything that relies on `all` arrays and them being "live" to all use map iterators and `entries()` is _maaybe_ possible, but a massive amount of work to avoid the possibility of accidentally updating only one of them in mutations.js.

The rest is a separate issue, unrelated to there being 2 separate references in cache.list/cache.map...

> but addObject wont account for the possibility that a different proxy object with the same data is already in the list.

This can't ever be allowed to be possible.  If it's happening (and you're saying it is, so I believe you), it's a 2nd bug.

That should be fixed by not ending up having 2 copies in the first place; trying to find equivalence between them is too late.  If there's multiple instances representing the same resource, one of those is a copy that's not going to get updates when a change event comes in and updates it.

> There's a fair chance that the resource already has a different proxy in the list because both the socket and the resource-instance save() function are trying to add one around the same time.

Have you seen this specifically happening for sure?  There's more async in here than in the ember client so I can see it being possible...

The copy from save() needs to be the one that "wins" and gets into the store first (using the `if (existing )` branch).

Then the change event comes in and loads some more data (which in this case is exactly the same data...), using the `else if ( entry )` branch because that entry was already found.

If the change event gets load()ed first, then the instance that is being `save()`-d is no longer going to become the model that gets persisted into the store and it won't get future updates.  (which is what the `existing` thing was added for to begin with.)

The 2.x client has a couple other differences which make this impossible-ish:
  - The model save() is what actually causes the equivalent of load() to happen (with less `async`'s in between)
  - The subscribe client writes all changes to a queue and flushes those changes to the actual store once a second to limit recomputes/repaints, (which dashboard should probably be doing too)

Reply:
@vincent99 I don't believe this is actually my issue. It looks like I just finally tracked down the issue and it's related to the store storing and returning dead arrays. I'll update my PR and add more commentary there. I think it will actually fix a number of the other refreshing issues we currently have.

Reply:
>Have you seen this specifically happening for sure? There's more async in here than in the ember client so I can see it being possible...

>The copy from save() needs to be the one that "wins" and gets into the store first (using the if (existing ) branch).

This is what I'm seeing (sometimes) when I include 
`addObject(cache.list, entry);`
in the first block in load() mutation. The list will then contain duplicate entries and the resource list view reflects that. 

I'm going to remove this issue from this PR and make a separate one for it so I can get the other secret-specific bug fixes here merged.